src/Opaleye/Internal/HaskellDB/Sql/Default.hs

```diff
@@ -17,20 +17,23 @@ import qualified Data.ByteString.Char8 as BS8
 import qualified Data.ByteString.Base16 as Base16
 import qualified Data.List.NonEmpty as NEL
 
-mkSqlGenerator :: SqlGenerator -> SqlGenerator
-mkSqlGenerator gen = SqlGenerator
+mkSqlGenerator :: SqlGenerator -> Maybe OnConflict -> SqlGenerator
+mkSqlGenerator gen conflict = SqlGenerator
```

Comment:
This seems weird to me since it's baking in the behavior of the `on conflict` strategy in to SQL generation. It seems like a SQL generator should generate for as many cases as possible, and I can imagine cases in which I would want to fail on conflicts in one table but not another. So I think this part of the design should be bubbled up farther ([reference](https://github.com/tomjaguarpaw/haskell-opaleye/pull/384/files#diff-170a1d73c68b44c6ced1062ece0f2b92R16)). Does that make sense?

That's really the only thing here I see that's different, but I really appreciate the work being done here!

Reply:
Your instincts are right.  This has been addressed in https://github.com/tomjaguarpaw/haskell-opaleye/tree/upsert which is the latest version of this PR.